doc/rest_api/snippet_sharing/api-description.apib

```diff
@@ -0,0 +1,587 @@
+
+        + password: `a_new-secureP4ssw0rd` (string, optional)
+
+            The new password for the user.
+
+            Regex: ^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)[a-zA-Z\d]{8,}$
+        
+        + rank: `3` (enum[number], optional)
+        
+            The new rank which should be set for the user. Setting this parameter does only result in a change if the currently authenticated user has the permission to change ranks.
+
+            Regex: ^[1-3]$
+        
+            + Members
+                + `1` - Normal user
+                + `2` - Moderator
+                + `3` - Admin
+            
++ Response 200 (application/json)
+    + Attributes (object)
+        + message (string) - Detailed response information, normally success message
+        + i18n (string) - A unique token representing above message; can be used for internationalization in frontends
+
+            A comprehensive list of possible messages:
+            - `USER_UPDATED_SUCCESSFULLY` if the changes to the user have been stored successfully
+
++ Response 400 (application/json)
+    + Attributes (object)
+        + status (string) - Textual description of the HTTP status
+        + message (string) - Detailed error information, e.g. hint about malformed request
+        + i18n (string) - A unique token representing above error description message; can be used for internationalization in frontends
+
+            A comprehensive list of possible errors:
+            - `USER_UPDATE_INVALID_PASSWORD` if the supplied `password` does not satisfy the given criteria (regex: ^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)[a-zA-Z\d]{8,}$)
+            - `USER_UPDATE_INVALID_EMAIL` if the supplied `email` is not a valid email
+            - `USER_UPDATE_INVALID_RANK` if the supplied `rank` is not a valid rank (regex: ^[1-3]$)
+
++ Response 401 (application/json)
+    + Attributes (object)
+        + status (string) - Textual description of the HTTP status
+        + message (string) - Detailed error information, e.g. hint about malformed request
+        + i18n (string) - A unique token representing above error description message; can be used for internationalization in frontends
+
+            A comprehensive list of possible errors:
+            - `NEED_AUTHENTICATION` if no valid session token has been supplied or the user for whom the token has been created does not exist (anymore)
+            - `USER_INSUFFICIENT_PERMISSIONS` if the currently authenticated user has insufficient permissions (due to his rank) to perform this action
+
+
+
+# Database endpoint [/database{?offset}{?rows}{?sort}{?sortby}{?filter}{?filterby}{?raw}]
+
+## List database entries [GET /database/{organization}/{application}/{scope}{?offset}{?rows}{?sort}{?sortby}{?filter}{?filterby}]
+
+Returns a list of database entries. The result can be adjusted by additional, but optional parameters.
+
++ Parameters
+    + organization (string, optional) - Only search within a given `organization` for entries
```

Comment:
If you want to search for an application from several organization, you can use the filter provided by the optional parameters. Although I got to say that this sounds terribly weird, like "the same application of different organizations".

> /apache/tomcat/common/some-snippet
> /org.apache/tomcat/common/some-other-snippet

You'd find both (and maybe more) by using either of the following requests:

> GET http://the-api.com/database/
> GET http://the-api.com/database/?filter=tomcat
> GET http://the-api.com/database/?filter=tomcat&filterby=key

If you provide a more specific URI (with an organization, an application and maybe even a scope), then this will only lead to a pre-filtering before the other options. If you leave the path empty, you can search all keys. The following should theoretically work as well with my design and also deliver both of above results:

> GET http://the-api.com/database/?filter=apache%2Ftomcat%2F&filterby=key
> ... which is the encoded form of ...
> GET http://the-api.com/database/?filter=apache/tomcat/&filterby=key